Speedup startup and toolbox operations

mvdbeek opened the pull request:
@bgruening mentioned slow toolbox reloading in gitter a while back, so I had a look at what where major things that could be sped up.

The most effective speedup for reloading the toolbox on my dev instance is rebuilding the search index only for the tools that have changed since the toolbox reloads. The speedup varies with how many tools are installed of course, for me this went from ~2-3 seconds to 0.4 seconds.

In terms of startup times the most effective change is caching the ToolShedRepositories for 1 second in 1b6da123f82e2e603f173aeb1b81ef0124ce6276 and always fetching the ToolVersion objects from the cache  in 4d5dd2d. This should be long enough to be able to load the complete toolbox with just a single emitted db query. This reduced the time to build the full galaxy app object from ~ 7.7 seconds to ~ 4.7 seconds. The more tools you have the better the speedup, I guess.

Data table reloading was also quite slow. It turns out there were hundreds of `os.path.exists()` calls, that even on my SSD were taking quite a while to finish. I addressed that in 12f7dcaf4eec0f20b3400783fb9ce1532401f934

I think this may even speedup other areas in galaxy, just running the unittests went from 140 seconds to 80 seconds locally. I guess everything that accesses a bunch of tools sequentially will benefit from this.

mvdbeek committed:
Use floating point to log toolox reload times

mvdbeek committed:
Add timing information for Data Manager reload

mvdbeek committed:
Add a short-lived (1s) ToolShedRepositoryCache

This will speed up repeated ToolShedRepositories lookups, like when
building the Toolbox for the first time, or when loading a
shed_tool_conf.xml file with many deactivated tools.

mvdbeek committed:
Speed up tool-data table reloads by doing a os.walk and caching the result for 1s instead of repeatedly calling os.path.exists.

mvdbeek committed:
Only index changed tools when re-indexing the toolbox

This speeds up repeated toolbox reloads enourmously (from 1 second on my
test-instance to 0.1 seconds).

mvdbeek committed:
Always use cache to fetch tool_version objects

mvdbeek committed:
Use cElementTree for light xml parsing speedup

mvdbeek committed:
Update ToolVersionCache when adding a new ToolShedLineage

mvdbeek committed:
Move tool_cache.reset_status() calls to reload toolbox/data managers, so that we can use information about whether tools are new during reloads

mvdbeek committed:
Make sure re-indexing the toolbox occurs on the new toolbox

mvdbeek committed:
Remove `replace_existing` option for data managers

If a data manager is defined twice for the same data manager id in the
shed_tool_conf.xml the first data manager will be loaded. This is not a
problem when updating a data manager through the toolshed, because the
outdated data manager will be removed from the tool cache and a new
toolbox is being built.

mvdbeek committed:
Reset the tool cache once all tools are loaded after starting up galaxy

mvdbeek committed:
Reduce logging during toolbox reloads

We don't log loading of tools that have not changed when reloading the
toolbox.

mvdbeek committed:
Fix unittests that access app.tool_cache

mvdbeek committed:
Log startup time

nsoranzo commented on lib/galaxy/queue_worker.py:
```diff
@@ -105,27 +108,33 @@ def _get_new_toolbox(app):
+    app.toolbox = new_toolbox
+    app.reindex_tool_search()
+    app.tool_cache.reset_status()
     return new_toolbox
```
You can remove the `return` statement.

nsoranzo commented on lib/galaxy/tools/data/__init__.py:
```diff
@@ -26,6 +27,31 @@
+        if time.time() - self.update_time > 1:
+            content = os.walk(self.tool_data_path)
+            try:
+                self._tool_data_path_files = set([os.path.join(dirpath, fn) for dirpath, _, fn_list in content for fn in fn_list if fn and fn.endswith('.loc') or fn.endswith('.loc.sample')])
```
You don't need the squared parentheses inside `set()`, i.e. you can just write `set(os.path.join...or fn.endswith('.loc.sample'))`

mvdbeek replied:
That's a nice one, I didn't know about that!

nsoranzo replied:
Yeah, you use a generator instead of creating the list first and then converting it to a set.

nsoranzo commented on lib/galaxy/tools/data/__init__.py:
```diff
@@ -26,6 +27,31 @@
+        if path in self.tool_data_path_files:
+            return True
+        else:
+            return False
```
```python
return path in self.tool_data_path_files
```
?

jmchilton committed:
Re-use ExecutionTimer abstractions for consistent log patterns.

jmchilton commented:
A few things:

- > This reduced the time to build the full galaxy app object from ~ 7.7 seconds to ~ 4.7 seconds.

  Wow - that is amazing - very nice!

- The API test are indicating there is something wrong with the data table loading I think.
- I opened a PR against your branch to normalize the log messages with other time tracking I do for various tool stuff - I use a fixed pattern to parse out timing information - there are some scripts in Galaxy to parse this stuff out. I think the pattern makes the log statements in Python look a little nicer as well.

mvdbeek commented:
Thanks @jmchilton @nsoranzo indeed, there are still some things left to be worked out, I just wanted to get some results from jenkins (your early feedback is great, of course!) ... somehow the toolshed tests are failing even on dev. Probably OSX to blame.

mvdbeek committed:
Fix cElementTree imports

mvdbeek committed:
Fix import order in lib/galaxy/tools/data/__init__.py

mvdbeek committed:
Add possibility explicitly rebuild ToolShedRepositoryCache

mvdbeek committed:
Fix path checking for files that are not in the tool_data_path directory

mvdbeek committed:
Use generator instead of list comprehenesion followed converting it to a set. (Thanks @nsoranzo)

mvdbeek committed:
Don't return toolbox (thx @nsoranzo)

mvdbeek committed:
Use ExecutionTimer abstraction to report startup time

mvdbeek committed:
Remove unused time import

mvdbeek committed:
Be more defesive when accessing app.tool_cache

mvdbeek committed:
Rebuild repo cache when deciding if repository should be created.

martenson commented on lib/galaxy/tools/search/__init__.py:
```diff
@@ -57,42 +57,57 @@ def build_index( self, index_help=True ):
+        return add_doc_kwds
+
+    def update_index(self, tool_cache, index_help=True):
+        """Use `tool_cache` to determine which tools need indexing and which tools should be expired."""
```
very nice!

mvdbeek committed:
Avoid having tools in tool_cache marked as new when they have rapidly shifted from new to expired.

mvdbeek committed:
Don't crash if a tool has not been added to the cache properly

mvdbeek committed:
Refresh tool_data_path_files if we know that a new table has been written

mvdbeek committed:
Reference old_toolbox earlier, since reloading can happen in a different thread, which will lead to a hang when waiting for another toolbox reload

mvdbeek committed:
Always refresh tool_shed_repository_cache before fetching repositories by ids (since that is only called from admin_toolshed.py and installed_repository_metadata_manager.py)

mvdbeek committed:
Fix import order in lib/galaxy/config.py

mvdbeek committed:
Allow looking up repositories by repository_id via ToolShedRepositoryCache

mvdbeek committed:
Refresh TSR cache when needed

mvdbeek committed:
Restore deactivate/activate functionality

mvdbeek committed:
Revert "Fix cElementTree imports"

This reverts commit 6d365f03e888ef12527105119f0a7adebae23b95.

mvdbeek committed:
Revert "Use cElementTree for light xml parsing speedup"

This reverts commit 556469f5171ef9897a5c40de9d661d2d4f4deb86.

mvdbeek committed:
Use ExecutionTimer

mvdbeek committed:
Use tool_shed_repository_cache in populate_tool_shed_info

mvdbeek committed:
Replace getchildren(), which has been deprecated (https://docs.python.org/2/library/xml.etree.elementtree.html#xml.etree.ElementTree.Element.getchildren)

mvdbeek committed:
Write out the datatype registry only once

mvdbeek committed:
Build toolbox search in a local control task

mvdbeek committed:
Don't fail tests without internet connectivity

mvdbeek committed:
Write out datatype registry when loading set_metadata tool

mvdbeek committed:
Keep cache thread-local

mvdbeek committed:
Use current session

mvdbeek committed:
Fix thread local cache use

mvdbeek committed:
Send control task as postfork function

mvdbeek committed:
Work around InstanceDetached errors

mvdbeek committed:
Skip time-based reloading, rely on detecting DetachedInstanceErrors to trigger cache rebuild

mvdbeek committed:
Attempt to speedup proprietary datatype loading

mvdbeek committed:
Remove operations from custom_datatype_manager

Updating the upload file formats is handled when loading the datatype
into the registry, and reloading the upload tool seems unnecessary and
insufficient, since this would only affect the current web handler. In
any case removed and added datatypes are immediately propagated to the
upload tool.

mvdbeek commented:
Thanks @jmchilton ... I think I'm close to the optimum for the things that I can improve without real-life feedback. I've brought the startup time down from 38 +/- 4 seconds on my server to 17+/- 2 seconds (the server does not yet have huge amount of tools installed, it's running in docker and some of the things are stored on a relatively slow network drive ...  YMMV). Profiling shows me that most of the time is being spent now in parsing xml, so apart from moving everything to yaml or getting cElementTree to work I don't think there is much left that I can do.

For the toolbox reloading part I strongly suspect that @bgruening's reloading slowness was due to deactivated or outdated entries in the shed_tool_conf.xml file. Those entries would never enter the tool_cache, so every single one of them was triggering a new db query. I hope I solved this with the ToolShedRepository cache. In any case I added a few timing statement here and there, so that we have a better idea of what could be slow.

jmchilton commented:
Opps sorry about closing that - I meant to cancel my half written comment not close your PR.

martenson reviewed (approved):
I have 91 tools locally and this PR:

Increases my startup toolbox indexing time from ~900ms to ~1.4s. 
Decreases my tool installation index rebuilding down to 33ms from ~700ms.
Leaves my startup time at ~8s.

Overall very nice improvement on the installation rebuild time.

@mvdbeek Can we somehow clean the async queue for index rebuilding if Galaxy goes down? It is useless at that point.

mvdbeek commented:
> Leaves my startup time at ~8s.

Hmm, that's a bit surprising, because that means the startup actually became slower, since the toolbox indexing doesn't count towards startup time anymore, or did you include the indexing in the timing?

Are you using sqlite locally ? It's possible that the ToolShedRepositoryCache doesn't do much for you if you are not using sqlite.

> Decreases my tool installation index rebuilding down to 33ms from ~700ms.

That was the primary objective, so hooray :)

> @mvdbeek Can we somehow clean the async queue for index rebuilding if Galaxy goes down? It is useless at that point.

Did you actually manage to queue up indexing form a previous restart?
I think we can check that the toolbox had been reloaded at least once before executing the task.
Which is not the same as cleaning the queue. @dannon do you have a suggestion of how to do that?

martenson commented:
@mvdbeek I have postgres on a very quick machine with ssd.

I included the index in timing. So ~6.7 without I guess.

And yeah, after one restart I had three queued up indexing tasks.

mvdbeek commented:
> And yeah, after one restart I had three queued up indexing tasks.

OK, we should definitely address that before merging. And I have seen an instance on my server where the search index had disappeared (no tools found by typing in the search box) until I triggered a toolbox reload.

martenson commented:
@mvdbeek Did I get it right at the call, that the queueing of indexing tasks has been removed from this and will be an independent PR? Do we want to remove WIP from this then?

mvdbeek commented:
No, I still have to do that.

mvdbeek committed:
Build search index at startup, rebuild repository cache during toolbox or datamanager reload

mvdbeek commented:
OK, I am building the toolbox search index during startup as we used to (https://github.com/galaxyproject/galaxy/pull/3909/commits/af156a8f7496dc9a9e70d3992f227c626aec7fa2), because doing that via the task queue didn't work on my production instance, which would start up with an empty tool search and only fill the search once the toolbox had been reloaded once. I am keeping track of how many times the toolbox had been reloaded, and reindex the search only if it has been rebuild less often than the toolbox. This way we ignore queued up rebuild requests. I think this is now ready for  a final round of review (ping @martenson).

martenson reviewed (approved):
Thank you Marius, again, a fabulous contribution! 💯
